ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+        1 != fread_function(&(pMMBDXP->day), 1, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (1 != fread_function(&nRecords, 4, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (1 != fread_function(&offset_primera_fitxa, 2, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    pMMBDXP->year = (short)(1900 + variable_byte);
+reintenta_lectura_per_si_error_CreaCampBD_XP:
+
+    if (n_queixes_estructura_incorrecta > 0)
+    {
+        if (!MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+        {
+            offset_fals = offset_primera_fitxa;
+            if ((offset_primera_fitxa - 1) % 32)
+            {
+                for (offset_fals = (offset_primera_fitxa - 1);
+                     !((offset_fals - 1) % 32); offset_fals--)
+                    ;
+            }
+        }
+    }
+    else
+        offset_reintent = ftell_function(pf);
+
+    if (1 != fread_function(&ushort, 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_1), 2, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->transaction_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->encryption_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->dbf_on_a_LAN), 12, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->nRecords, &nRecords, 4);
+        memcpy(((char *)&pMMBDXP->nRecords) + 4, &pMMBDXP->dbf_on_a_LAN, 4);
+    }
+    else
+        pMMBDXP->nRecords = nRecords;
+
+    if (1 != fread_function(&(pMMBDXP->MDX_flag), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->CharSet), 1, 1, pf) ||
+        1 != fread_function(&(pMMBDXP->reserved_2), 2, 1, pf))
+    {
+        fclose_function(pf);
+        return 1;
+    }
+
+    // Checking for a cpg file
+    if (pMMBDXP->CharSet == 0)
+    {
+        FILE_TYPE *f_cpg;
+
+        strcpy(cpg_file, pMMBDXP->szFileName);
+        strcpy(cpg_file, reset_extension(cpg_file, "cpg"));
+        f_cpg = fopen_function(cpg_file, "r");
+        if (f_cpg)
+        {
+            char *p;
+            size_t read_bytes;
+            fseek_function(f_cpg, 0L, SEEK_SET);
+            if (11 > (read_bytes = fread_function(local_message, 1, 10, f_cpg)))
+            {
+                local_message[read_bytes] = '\0';
+                p = strstr(local_message, "UTF-8");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_UTF8_DBF;
+                p = strstr(local_message, "UTF8");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_UTF8_DBF;
+                p = strstr(local_message, "ISO-8859-1");
+                if (p)
+                    pMMBDXP->CharSet = MM_JOC_CARAC_ANSI_DBASE;
+            }
+            fclose_function(f_cpg);
+        }
+    }
+    if (MM_ES_DBF_ESTESA(pMMBDXP->dbf_version))
+    {
+        memcpy(&pMMBDXP->FirstRecordOffset, &offset_primera_fitxa, 2);
```

Comment:
It is a bit scary to copy only 2 bytes in a field that is a GInt32. It would be better to copy them in a temporary  unsigned short, and then reassemble the full GInt32 from the 2 shorts

Reply:
Yes it's the idea of my boss that wanted to use some unused bytes from DBASEIV format to save some information.
It's explained at this document that we reviewed and translated to english just for GDAL project ;)
https://www.miramon.cat/new_note/eng/notes/DBF_estesa.pdf

I'll try to do the change you want. (It made me laugh the word "scary")

Reply:
This is more understandable?
unsigned short FirstRecordOffsetPart1;
        unsigned short FirstRecordOffsetPart2;

        memcpy(&FirstRecordOffsetPart1, &offset_primera_fitxa, 2);
        memcpy(&FirstRecordOffsetPart2, &pMMBDXP->reserved_2, 2);

        pMMBDXP->FirstRecordOffset = (int)FirstRecordOffsetPart1 << 16 | FirstRecordOffsetPart2;

Reply:
instead of int MM_FIRST_RECORD_OFFSET_TYPE or GInt32 (it's the same) at the end

Reply:
opposite way. let me do it again in a better way:

unsigned short FirstRecordOffsetPart1;
        unsigned short FirstRecordOffsetPart2;

        memcpy(&FirstRecordOffsetPart1, &offset_primera_fitxa, 2);
        memcpy(&FirstRecordOffsetPart2, &pMMBDXP->reserved_2, 2);

        pMMBDXP->FirstRecordOffset =
            ((GUInt32)FirstRecordOffsetPart2 << 16) | FirstRecordOffsetPart1;

Reply:
If nothing to say, for me it's closed

Reply:
FirstRecordOffsetPart1 ==> FirstRecordOffsetLow16Bits
FirstRecordOffsetPart2 ==> FirstRecordOffsetHigh16Bits

Reply:
Done!